extensions/amp-react-img/0.1/amp-react-img.js

```diff
@@ -0,0 +1,324 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Layout, isLayoutSizeDefined} from '../../../src/layout';
+import {Services} from '../../../src/services';
+import {devAssert} from '../../../src/log';
+import {requireExternal} from '../../../src/module';
+
+/**
+ * @param {!Object<string, *>} props
+ * @param {!Array<string>} keys
+ * @return {!Object<string, *>}
+ */
+function pick(props, keys) {
+  const out = {};
+  for (let i = 0; i < keys.length; i++) {
+    const key = keys[i];
+    const value = props[key];
+    if (value !== undefined) {
+      out[key] = value;
+    }
+  }
+
+  return out;
+}
+
+/**
+ * @return {boolean}
+ */
+function testSrcsetSupport() {
+  const support = 'srcset' in new Image();
+  testSrcsetSupport = () => support;
+  return support;
+}
+
+/**
+ * @param {string|undefined} src
+ * @param {string|undefined} srcset
+ * @return {string|undefined}
+ */
+function guaranteeSrcForSrcsetUnsupportedBrowsers(src, srcset) {
+  if (src !== undefined || testSrcsetSupport()) {
+    return src;
+  }
+  const match = /\S+/.exec(srcset);
+  return match ? match[0] : undefined;
+}
+
+const ATTRIBUTES_TO_PROPAGATE = [
+  'alt',
+  'title',
+  'referrerpolicy',
+  'aria-label',
+  'aria-describedby',
+  'aria-labelledby',
+  'srcset',
+  'src',
+  'sizes',
+];
+
+AMP.extension('amp-react-img', '0.1', AMP => {
+  const react = requireExternal('react');
+  const {render} = requireExternal('react-dom');
+
+  /**
+   * We'll implement all our new extensions as React/Preact Components (TBD).
+   * They're true Components, not AmpElements/Amp.BaseElements.
+   */
+  class AmpImg extends react.Component {
+    /**
+     * @param {!Object} props
+     */
+    constructor(props) {
+      super(props);
+
+      this.state = {
+        prerender: true,
+        element: props['i-amphtml-element'],
+        layoutWidth: 0,
+      };
+      delete props['i-amphtml-element'];
+
+      /** @private {boolean} */
+      this.prerenderAllowed_ = !!props['noprerender'];
+
+      /** @private {number} */
+      this.currentSizesWidth_ = 0;
+
+      /** @private {string|undefined} */
+      this.currentSizes_ = undefined;
+    }
+
+    /**
+     * @return {*}
+     */
+    render() {
+      const props = pick(this.props, ATTRIBUTES_TO_PROPAGATE);
+
+      const {id} = this.props;
+      if (id) {
+        props['amp-img-id'] = id;
+      }
+      props['src'] = guaranteeSrcForSrcsetUnsupportedBrowsers(
+        props['src'],
+        props['srcset']
+      );
+      props['sizes'] = this.maybeGenerateSizes_(props['sizes']);
+      props['decoding'] = 'async';
+
+      const {prerender} = this.state;
+      if (prerender && !this.prerenderAllowed_) {
```

Comment:
I'm curious, why is `this.state. prerender ` and `this.prerenderAllowed_` needed? Why not directly check the prop value for `noprerender` to determine if `src` and `srcset` should be deleted?

```js
if (!this.props.noprerender) {
  delete props['src'];
  delete props['srcset'];
}
```

Reply:
Hi!

I think we actually meant to do `!prerender || !this.prerenderAllowed`. Note that we have to different sources here, one from `this.state.prerender` (which is outside-of-React props being passed to React) and `props.noprerender` (which is both the DOM-React and React-React possible values). We use the `state` to control from the AMP framework, and props to allow the publisher to control behavior.

Also, this PR is stagnant. We started a new attempt at https://github.com/ampproject/amphtml/pull/25969. Feel free to follow along!

Reply:
Thanks for the info!